File: 김소희/택배 배송.py
import heapq
from collections import defaultdict

N, M = map(int, input().split())

graph = defaultdict(list)
for _ in range(M):
    a, b, cost = map(int, input().split())
    graph[a].append((cost, b))
    graph[b].append((cost, a))

pq = [(0, 1)]
costL = [float('inf')] * (N+1)
costL[1] = 0

while pq:
    cost, node = heapq.heappop(pq)
    if costL[node] < cost:
        continue

    costL[node] = cost
    for w, next in graph[node]:
        newCost = costL[node] + w
        if costL[next] > newCost:
            costL[next] = newCost
            heapq.heappush(pq, (costL[next], next))
            
print(costL[N])

File: 김소희/택배 배송.java
import java.io.*;
import java.util.*;

public class Main {

    static class Edge {
        int to;
        int cost;

        Edge(int to, int cost) {
            this.to = to;
            this.cost = cost;
        }
    }

    public static void main(String[] args) throws Exception {

        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st = new StringTokenizer(br.readLine());

        int N = Integer.parseInt(st.nextToken());
        int M = Integer.parseInt(st.nextToken());

        List<Edge>[] graph = new ArrayList[N + 1];
        for (int i = 1; i <= N; i++) {
            graph[i] = new ArrayList<>();
        }

        for (int i = 0; i < M; i++) {
            st = new StringTokenizer(br.readLine());
            int a = Integer.parseInt(st.nextToken());
            int b = Integer.parseInt(st.nextToken());
            int cost = Integer.parseInt(st.nextToken());

            graph[a].add(new Edge(b, cost));
            graph[b].add(new Edge(a, cost));
        }

        int[] dist = new int[N + 1];
        Arrays.fill(dist, Integer.MAX_VALUE);
        dist[1] = 0;

        PriorityQueue<int[]> pq =
                new PriorityQueue<>(Comparator.comparingInt(o -> o[0]));
        pq.offer(new int[]{0, 1}); // (비용, 노드)

        while (!pq.isEmpty()) {
            int[] cur = pq.poll();
            int cost = cur[0];
            int node = cur[1];

            if (dist[node] < cost) continue;

            for (Edge e : graph[node]) {
                int next = e.to;
                int newCost = dist[node] + e.cost;

                if (newCost < dist[next]) {
                    dist[next] = newCost;
                    pq.offer(new int[]{newCost, next});
                }
            }
        }

        System.out.println(dist[N]);
    }
}
